stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -125,275 +125,185 @@ end
+#!
+#! Inputs:
+#!   Operand stack: [ptr, PK, ...]
+#!   Advice stack:  [tau0, tau1, h_0, ..., h_511, s2_0, ..., s2_511, pi_0, ..., pi_1022, ...]
+#! Outputs:
+#!   Operand stack: []
+#!   Advice stack:  []
+#!
+#! Cycles: 4430
+export.load_h_s2_and_product.12
+    # We will store in local addresses:
+    # 1. 0..4 the public key,
+    # 2. 4..8 the inverse of the evaluation point tau as [tau_inv0, tau_inv1, 0, 0],
+    # 3. 8..12 the evaluation point tau as [tau0, tau1, 0, 0],
+    
+    # 1) Set up the stack for loading the coefficients of the polynomials, evaluating and hashing them
+    
+    ## a) Set up the accumulator for `horner_eval_base` and the memory pointers
+    push.0.0
+    locaddr.4
+    movup.3
 
-    # 1) Store PK for later comparison
-    movdn.4
+    ## b) Save PK to later compare it with the hash of the h polynomial
+    swapw
     loc_storew.0
+    # => [PK, ptr, alpha_inv, acc1, acc0, ...]
 
-    # 2) Prepare stack and load h polynomial. We also range check the coefficients of h.
+    ## c) Prepare the capacity portion of the state of the hasher
     padw swapw
-    padw
+    # => [Y, 0, 0, 0, 0, ptr, alpha_inv, acc1, acc0, ...], where Y is a "garbage" word
+
+    ## d) Load the evaluation point tau from the advice tape, compute its inverse and save both
+    ## Note that we will evaluate the polynomials at `tau_inv = tau^{-1}` as we are loading the coefficients
+    ## in the normal order i.e., not in reversed order required for Horner evaluation.
+    ## This means that the equality we will be checking is 
+    ## pi(tau_inv) * tau^1023 == h(tau_inv) * tau^511 * s2(tau_inv) * tau^511
+    ##
+    ## which simplifies to
+    ##
+    ## pi(tau_inv) * tau == h(tau_inv) * s2(tau_inv)
+    adv_push.2 
+    dup.1 dup.1 ext2inv
+    push.0.0
+    loc_storew.4
+
+    ## e) Store the random evaluation point. We will later check that it is the correct one.
+    movup.2 drop
+    movup.2 drop
+    loc_storew.8
+    # => [Y, Y, 0, 0, 0, 0, ptr, alpha_inv, acc1, acc0, ...]
+
+    # 2) Load the coefficients of the h polynomial and evaluate them at tau_inv
     repeat.64
         adv_pipe
 
+        # check that all coefficients are less than Falcon prime
         dupw.1
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
 
         dupw
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
+
+        horner_eval_base
 
         hperm
     end
```

Comment:
Could we add stack state after this loop?

Reply:
done